### PR TITLE
KAFKA-5673: refactor KeyValueStore hierarchy to make MeteredKeyValueStore outermost

### PR DESCRIPTION
**Pull request by dguy**
refactor StateStoreSuppliers such that a `MeteredKeyValueStore`  is the outermost store.

### CODE REVIEW

**Comment by dguy** [comment]
@guozhangwang @bbejeck @mjsax  - one thing i'm not sure about is if we should make every store use a `MeteredKeyValueBytesStore`, i.e, merge it back into one store rather than having two. The downside is that for InMemory stores we will still be serializing/deserializing - thoughts?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6396/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6411/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6426/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6441/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6430/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6445/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6446/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6431/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6545/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6531/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by dguy** [comment]
@guozhangwang @bbejeck @enothereska can i please get a review?

**Review by enothereska — APPROVED** [review]
LGTM. I think it's ok to have the 2 versions for now to avoid the cost for in-memory stores.

**Comment by bbejeck** [comment]
@dguy thanks for this.  LGTM.

I agree with @enothereska.   I think the second version is worth the trade-off for the in-memory stores.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java [hunk 1]
```diff
@@ -182,47 +177,43 @@ public synchronized long approximateNumEntries() {
...
-    public synchronized V delete(final K key) {
+    public synchronized byte[] delete(final Bytes key) {
         validateStoreOpen();
         Objects.requireNonNull(key);
-        final byte[] rawKey = serdes.rawKey(key);
-        final Bytes bytesKey = Bytes.wrap(rawKey);
-        final V v = get(rawKey);
-        cache.delete(cacheName, bytesKey);
-        underlying.delete(bytesKey);
+        final byte[] v = getInternal(key);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java:214]: Nice catch.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InnerMeteredKeyValueStore.java [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+        }
+
+    }
+
+    @Override
+    public long approximateNumEntries() {
+        return inner.approximateNumEntries();
+    }
+
+    interface Action<V> {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InnerMeteredKeyValueStore.java]: This interface can be private?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueBytesStore.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        this.valueSerde = valueSerde;
+        innerMetered = new InnerMeteredKeyValueStore<>(inner, metricScope, new InnerMeteredKeyValueStore.TypeConverter<K, Bytes, V, byte[]>() {
+            @Override
+            public Bytes innerKey(final K key) {
+                return Bytes.wrap(serdes.rawKey(key));
+            }
+
+            @Override
+            public byte[] innerValue(final V value) {
+                return serdes.rawValue(value);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueBytesStore.java]: Not a comment: as we have concluded in https://issues.apache.org/jira/browse/KAFKA-4750 we should check for null here and skip. cc @evis.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InnerMeteredKeyValueStore.java [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+ * inner KeyValueStore implementation do not need to provide its own metrics collecting functionality.
+ *
+ * @param <K>
+ * @param <V>
+ */
+class InnerMeteredKeyValueStore<K, IK, V, IV> extends WrappedStateStore.AbstractStateStore implements KeyValueStore<K, V> {
+
+    private final KeyValueStore<IK, IV> inner;
+    private final String metricScope;
+    private final TypeConverter<K, IK, V, IV> typeConverter;
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InnerMeteredKeyValueStore.java]: We also need to explain a bit why we add a type converter at this layer of the store hierarchy.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueBytesStore.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.StateSerdes;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A Metered {@link KeyValueStore} wrapper that is used for recording operation metrics, and hence its
+ * inner KeyValueStore implementation do not need to provide its own metrics collecting functionality.
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueBytesStore.java:35]: Also include the serde functionality in java doc?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheKeyValueStoreIterator.java [hunk 6]
```diff
@@ -19,43 +19,37 @@
...
-import org.apache.kafka.streams.state.StateSerdes;
 
 /**
  * Merges two iterators. Assumes each of them is sorted by key
  *
- * @param <K>
- * @param <V>
  */
-class MergedSortedCacheKeyValueStoreIterator<K, V> extends AbstractMergedSortedCacheStoreIterator<K, Bytes, V, byte[]> {
+class MergedSortedCacheKeyValueStoreIterator extends AbstractMergedSortedCacheStoreIterator<Bytes, Bytes, byte[], byte[]> {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheKeyValueStoreIterator.java]: nit: rename to `MergedSortedCacheKeyValueBytesStoreIterator`?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueBytesStore.java [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+ *
+ * @param <K>
+ * @param <V>
+ */
+public class MeteredKeyValueBytesStore<K, V> extends WrappedStateStore.AbstractStateStore implements KeyValueStore<K, V> {
+
+    private final Serde<K> keySerde;
+    private final Serde<V> valueSerde;
+    private StateSerdes<K, V> serdes;
+    private InnerMeteredKeyValueStore<K, Bytes, V, byte[]> innerMetered;
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueBytesStore.java]: This field can be final as well.

**Review by guozhangwang — COMMENTED** [review]
I treat this PR as part of the KIP-182, as the current proposal only allows user to pass in `BytesStoreSupplier` either in DSL or PAPI to overridde the default storage engine. Here's my thoughts on @dguy 's question:

1. For in-memory store, they are less common use cases and the serde cost are worthwhile to avoid kicking memory management out of the door.
2. For serde cost in general, we still have one major optimization opportunity in the future to use some "Structured KStream / KTable" which are schema-aware and hence can map operators on the close-formed fields into byte array manipulations directly.
3. As for KIP-182, the current byte store APIs do not prevent extensions to support generic typed stores if such request becomes necessary in the future.

So I'm inclined to enforce the bytes store for in-memory kv or LRU as well.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InnerMeteredKeyValueStore.java (hunk 4 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/InnerMeteredKeyValueStore.java]: Added to the interface def

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6700/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6685/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by bbejeck** [comment]
> For in-memory store, they are less common use cases and the serde cost are worthwhile to avoid kicking memory management out of the door.

I definitely can see your point.

However, even if in-memory store is less common,  I'm hesitant to force users into serialization who want the performance of an in-memory store. 
As for memory management we could add documentation stating that using this approach puts more responsibility on the implementors to ensure the store is used properly.  

>As for KIP-182, the current byte store APIs do not prevent extensions to support generic typed stores if such request becomes necessary in the future.

True, but I don't want to be too prescriptive and would prefer to leave the option of an in-memory store in place, without having to write a custom store. 

Either way I think it's ok to enforce bytes for in-memory or LRU stores, but IMHO the optimal solution is to have an option of in-memory stores with out bytes and let implementors decide.

Going forward I think one of our best options is to fully explore the "Structured KStream / KTable" you mention in point two above.

**Comment by dguy** [comment]
@bbejeck with KIP-182 you can have in In Memory store, but it will always be `<Bytes, byte[]>` as that is the only way we can support it without having many overloads and ugliness. That said, they can still use the existing `Stores` API and soon to be deprecated methods on `KStream` etc to have a non-serialzied `InMemoryStore`

It is less than perfect, but it means that anyone can write a custom store and have it auto-magically wrapped with change logging, caching, metering, anything else we come up with (which isn't the case today). I think the trade-off is reasonable as for the majority of cases people don't really care about having custom stores.
Yes, going forward it will be great if we can do the "Structured streaming" thing, then the perf hit goes away in some respect

**Comment by guozhangwang** [comment]
LGTM. Thanks.

**Pull request closed without merging**